src/opm/parser/eclipse/Parser/raw/RawKeyword.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+  Copyright 2013 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <stdexcept>
+#include <boost/algorithm/string.hpp>
+
+#include <opm/parser/eclipse/Parser/ParserKeyword.hpp>
+#include <opm/parser/eclipse/Utility/String.hpp>
+
+#include "RawConsts.hpp"
+#include "RawKeyword.hpp"
+#include "RawRecord.hpp"
+
+namespace Opm {
+
+namespace {
+
+     std::string keyword_name(const std::string& input_name) {
+         std::string name = boost::algorithm::trim_right_copy(input_name);
+         if (!ParserKeyword::validDeckName(name))
+            throw std::invalid_argument("Not a valid keyword:" + name);
+
+         if (name.size() > Opm::RawConsts::maxKeywordLength)
+            throw std::invalid_argument("Too long keyword:" + name);
+
+         if (name[0] == ' ')
+             throw std::invalid_argument("Illegal whitespace start of keyword:" + name);
+
+         return name;
+    }
+
+}
+
+
+    RawKeyword::RawKeyword(const std::string& name, const std::string& filename, std::size_t lineNR, bool raw_string, Raw::KeywordSizeEnum sizeType, std::size_t size_arg) :
+        m_name(keyword_name(name)),
+        m_filename(filename),
+        m_lineNR(lineNR),
+        raw_string_keyword(raw_string),
+        m_sizeType(sizeType)
+    {
+        printf("Creating %s size_type: %d:%ld \n", this->m_name.c_str(), static_cast<int>(sizeType), size_arg);
+        if (this->m_sizeType == Raw::FIXED) {
+            this->m_fixedSize = size_arg;
+            if (size_arg == 0) {
+                this->m_isFinished = true;
+                printf("Setting finished = True\n");
+            }
+        }
+
+        if (this->m_sizeType == Raw::TABLE_COLLECTION) {
+            if (size_arg == 0)
+                throw std::logic_error("Can not create TableCollection with zero tables.");
+            this->m_numTables = size_arg;
+        }
+
+        if (this->m_sizeType == Raw::SLASH_TERMINATED || this->m_sizeType == Raw::UNKNOWN) {
+            if (size_arg != 0)
+                throw std::logic_error("Must have size_arg == 0 for SLASH_TEMINATED and UNKNOWN");
+        }
+
+
+        if (this->m_sizeType == Raw::CODE) {
+            if (size_arg != 1)
+                throw std::logic_error("Must have size_arg == 1 for CODE");
```

Comment:
I think diagnostic would be more useful if it included the keyword name and, possibly, the location/line number.